extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java

```diff
@@ -94,34 +94,29 @@ public Result parse(MethodInfo methodInfo) {
 
         // handle 'Top' and 'First'
         Integer topCount = null;
-        int firstIndex = methodName.indexOf("First");
-        int topIndex = methodName.indexOf("Top");
-        if ((firstIndex != -1) || (topIndex != -1)) {
-            int effectiveFirstOrTopIndex = (topIndex > -1 && firstIndex > -1) ? Math.max(topIndex, firstIndex)
-                    : Math.max(firstIndex, topIndex);
-            // 'First' and 'Top' could be part of a field name, so we only conside them as part of a top query
-            // if they are found before 'By'
-            if (effectiveFirstOrTopIndex < byIndex) {
-                try {
-                    String topCountStr = methodName.substring(effectiveFirstOrTopIndex, byIndex)
-                            .replace("Top", "").replace("First", "");
-                    if (topCountStr.isEmpty()) {
-                        topCount = 1;
-                    } else {
-                        topCount = Integer.valueOf(topCountStr);
-                    }
-                } catch (Exception e) {
-                    throw new UnableToParseMethodException(
-                            "Unable to parse query with limiting results clause. Offending method is "
-                                    + repositoryMethodDescription + ".");
+        int minFirstOrTopIndex = Math.min(indexOfOrMaxValue(methodName, "First"), indexOfOrMaxValue(methodName, "Top"));
+        // 'First' and 'Top' could be part of a field name, so we only conside them as part of a top query
+        // if they are found before 'By'
+        if (minFirstOrTopIndex < byIndex) {
+            if (queryType != QueryType.SELECT) {
+                throw new UnableToParseMethodException(
+                        "When 'Top' or 'First' is specified, the query must be a find query. Offending method is "
+                                + repositoryMethodDescription + ".");
+            }
+            try {
+                String topCountStr = methodName.substring(minFirstOrTopIndex, byIndex)
+                        .replace("Top", "").replace("First", "");
+                if (topCountStr.isEmpty()) {
+                    topCount = 1;
+                } else {
+                    topCount = Integer.valueOf(topCountStr);
                 }
+            } catch (Exception e) {
+                throw new UnableToParseMethodException(
+                        "Unable to parse query with limiting results clause. Offending method is "
+                                + repositoryMethodDescription + ".");
             }
         }
-        if ((topCount != null) && (queryType != QueryType.SELECT)) {
```

Comment:
I moved this up into the `minFirstOrTopIndex < byIndex` block because that's the ony place where `topCount` is set to something != null.

Reply:
Sounds good